tools/bidder-bot/notifier/full.go

```diff
@@ -0,0 +1,146 @@
+package notifier
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const (
+	BlockDuration = 12 * time.Second
+)
+
+type FullNotifier struct {
+	logger               *slog.Logger
+	targetBlockNumChan   chan uint64
+	l1Client             L1Client
+	notifySecondsAhead   time.Duration
+	lastNotifiedBlockNum uint64
+	mu                   sync.Mutex
+}
+
+func NewFullNotifier(
+	logger *slog.Logger,
+	l1Client L1Client,
+	notifySecondsAhead time.Duration,
+	targetBlockNumChan chan uint64,
+) *FullNotifier {
+	return &FullNotifier{
+		logger:             logger,
+		l1Client:           l1Client,
+		notifySecondsAhead: notifySecondsAhead,
+		targetBlockNumChan: targetBlockNumChan,
+	}
+}
+
+type L1Client interface {
+	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
+}
+
+func (b *FullNotifier) Start(ctx context.Context) <-chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		headers := make(chan *types.Header)
+		sub, err := b.l1Client.SubscribeNewHead(ctx, headers)
+		if err != nil {
+			b.logger.Error("failed to subscribe to new heads", "error", err)
+			return
+		}
+
+		b.logger.Info("subscribed to new block headers")
+
+		for {
+			select {
+			case <-ctx.Done():
+				b.logger.Info("context done")
+				return
+
+			case err := <-sub.Err():
+				b.logger.Error("subscription error", "error", err)
+				sub, err = b.l1Client.SubscribeNewHead(ctx, headers)
+				if err != nil {
+					b.logger.Error("failed to resubscribe to new heads", "error", err)
+					return
+				}
+
+			case header := <-headers:
+				if err := b.HandleHeader(ctx, header); err != nil {
+					b.logger.Error("error handling header", "error", err)
+				}
+			}
+		}
+	}()
+	return done
+}
+
+func (b *FullNotifier) HandleHeader(ctx context.Context, header *types.Header) error {
+	targetBlockNum := header.Number.Uint64() + 1
+	currentBlockTime := time.Unix(int64(header.Time), 0)
+	nextBlockTime := currentBlockTime.Add(BlockDuration)
+	notificationTime := nextBlockTime.Add(-b.notifySecondsAhead)
+
+	if notificationTime.Before(time.Now()) {
+		if err := b.sendTargetBlockNotification(targetBlockNum); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	b.logger.Debug("scheduling notification",
+		"target_block_number", targetBlockNum,
+		"notification_time", notificationTime,
+		"expected_block_time", nextBlockTime)
+
+	b.scheduleNotification(ctx, targetBlockNum, notificationTime)
+	return nil
+}
+
+func (b *FullNotifier) scheduleNotification(ctx context.Context, targetBlockNum uint64, notificationTime time.Time) {
```

Comment:
There is again no need to schedule notification. You can just wait till time passes as your consumer is not going to get more blocks till time passes. The extra go  routine is not required.

Reply:
Sending the bid at a specific time during the block sealing process was a request from @iowar. He requested that bids are sent ~7 seconds before the block is sealed if I remember correctly. This is why we delay the notification. To remove this functionality, I'd first need confirmation from Iowar that it would be ok

Reply:
Simplified things in https://github.com/primev/mev-commit/commit/5ce105a78e93e2e88482b350621249a05573322c